pymongo/uri_parser.py

```diff
@@ -141,7 +149,18 @@ def parse_userinfo(userinfo):
     # No password is expected with GSSAPI authentication.
     if not user:
         raise InvalidURI("The empty string is not valid username.")
-    return unquote_plus(user), unquote_plus(passwd)
+
+    # If we cannot round-trip a value, it is not properly URI encoded.
+    for value_name, quoted_value in zip(
+            ('username', 'password',), (user, passwd,)):
+        quoted_value_no_sdelims = "".join(
+            [ch for ch in quoted_value if ch not in SUBDELIMS])
+        unquoted_value = unquote_plus(quoted_value_no_sdelims)
+        if not quoted_value_no_sdelims == quote_plus(unquoted_value):
+            raise InvalidURI("%r is not a valid %s" % (
+                quoted_value, value_name))
+
+    return unquote(user), unquote(passwd)
```

Comment:
Isn't this a breaking change? A plus sign (`+`) used to turn into a space (` `) and now it does not?

Reply:
It seems that way. See my comment https://github.com/mongodb/mongo-python-driver/pull/409/files#r261802812. It seems like I MUST make this change in order to get the test entitled "Subdelimiters in user/pass don't need escaping (MONGODB-CR)" to pass.

Reply:
I support this change (in PyMongo 4.0). We use (un)quote_plus because of a historical misunderstanding. Using just quote would have the side benefit of PyMongo matching the quoting rules supported by WinKerberos. We'll have to call this out boldly in the changelog for 4.0, however.